code/hash/Hacl.Hash.Definitions.fst

```diff
@@ -156,6 +156,43 @@ let hash_len (a: fixed_len_alg): n:size_t { v n = hash_length a } =
   | SHA3_384 -> 48ul
   | SHA3_512 -> 64ul
 
+/// Maximum input length, but fitting on a 64-bit integer (since the streaming
+/// module doesn't bother taking into account lengths that are greater than
+/// that). The comment previously was:
+///
+/// Note that we keep the total length at run-time, on 64 bits, but require that
+/// it abides by the size requirements for the smaller hashes -- we're not
+/// interested at this stage in having an agile type for lengths that would be
+/// up to 2^125 for SHA384/512.
+
+module U64 = FStar.UInt64
+
+inline_for_extraction noextract
+let max_input_len64 a: U64.(x:t { 0 < v x /\ v x `less_than_max_input_length` a }) =
```

Comment:
sadly we can't because the algorithm `a` is erased -- this is intentional, because the `hacl_keccak` works for *all* instances of keccak without doing any branching at run-time

(see line 120 of Hacl.Streaming.Keccak:

```
inline_for_extraction noextract
let hacl_keccak (a: G.erased alg): block alg =
  Block
    Erased
    stateful_keccak (* state *)
    (stateful_unused alg) (* key *)
    Lib.IntTypes.(x:size_t { v x > 0 }) (* output_length_t *)

    (fun _ -> 0xffffffffffffffffUL) (* max_input_len *)
```
)

I've therefore implemented one of your earlier suggestions:

```
    (fun _ -> Lib.IntTypes.(ones U64 PUB)) (* max_input_len *)
```

which is of course much more readable and better.

This also works:

```

    (fun _ ->
      [@inline_let]
      let max = Lib.IntTypes.(ones U64 PUB) in
      assert (forall (a: alg). Hacl.Hash.Definitions.max_input_len64 a == max);
      max) (* max_input_len *)
```